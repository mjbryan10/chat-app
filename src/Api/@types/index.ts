export interface User {
  id: number;
  name: string;
}

export interface LoginResponse extends User {
  isAuthenticated: boolean;
}