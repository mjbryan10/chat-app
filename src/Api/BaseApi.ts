export default class BaseApi {
  private readonly _baseUrl = 'http://assignment.bunq.com/'

  get baseUrl() {
    return this._baseUrl
  }
}