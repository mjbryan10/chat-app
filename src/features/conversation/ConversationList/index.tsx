import React, { useEffect } from 'react';
import * as S from './styles';
import { useSelector, useDispatch } from 'react-redux';
import {
   selectConversations,
   fetchConversations,
   clearConversations,
   selectConversationStatus,
} from '../conversationSlice';
import ConversationItem from '../ConversationItem';
import { selectLogin } from 'features/login/loginSlice';

const ConversationList = () => {
   const conversations = useSelector(selectConversations);
   const { id: currentUserId } = useSelector(selectLogin);
   const status = useSelector(selectConversationStatus);
   const dispatch = useDispatch();
   useEffect(() => {
      if (currentUserId) {
         dispatch(fetchConversations(currentUserId));
      }
      return () => {
         dispatch(clearConversations);
      };
   }, [currentUserId, dispatch]);
   return (
      <S.Container>
         <S.Header>Conversations</S.Header>
         {status === 'pending' ? 'loading...' : null}
         {currentUserId
            ? conversations.map((conversation) => {
                 const { conversationId } = conversation.conversation;
                 return (
                    <ConversationItem
                       conversation={conversation}
                       key={conversationId}
                       currentUserId={currentUserId}
                    />
                 );
              })
            : null}
      </S.Container>
   );
};

export default ConversationList;
