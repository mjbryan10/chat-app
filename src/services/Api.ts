const Api = {
  baseUrl: 'http://assignment.bunq.com'
}

export default Api;