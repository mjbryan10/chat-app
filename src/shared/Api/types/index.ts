export interface User {
   id: number;
   name: string;
}

export interface LoginResponse extends User {
   isAuthenticated: boolean;
}

export interface ConversationDetails {
   id: number | null;
   conversationId: number | null;
   is_owner: number;
   userid: number;
   status: number | null;
   lastseen: string | null;
   name: string | null;
   type: number | null;
}

export interface ConversationUser {
   id: number;
   conversationId: number;
   is_owner: number;
   userid: number;
   status: number | null;
   lastseen: string | null;
}

export interface Conversation {
   conversation: ConversationDetails;
   users: ConversationUser[];
}

export interface Message {
   id: number;
   senderId: number;
   message: string;
   timestamp: string;
   conversationid: number;
   status: number | null;
}

export type NewMessagesResponse = Message[] | { message: string };

export interface ApiPostResponse {
   id: number;
}

export const TIME_FORMAT = 'YYYY-MM-DD hh:mm:ss'