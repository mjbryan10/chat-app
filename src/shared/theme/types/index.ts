/**
 * Class for making an array of colors
 * 
 * 
 * Technique to avoid hard typing names into array:
 * @see https://stackoverflow.com/a/59806829/12873927
 */
class ColorSpectrumList {
   /**
    * 
    */
   constructor(
      readonly blue?: string,
      readonly red?: string,
      readonly green?: string,
      readonly purple?: string,
      readonly orange?: string,
      readonly pink?: string,
      readonly indigo?: string,
      readonly teal?: string,
      readonly yellow?: string,
      readonly cyan?: string,
   ){}
}
export type SpectrumColor = keyof ColorSpectrumList;
/**
 * An array of color names for use.
 */
export const colorSpectrumArray = Object.keys(new ColorSpectrumList()) as SpectrumColor[]; 
/**
 * Color Spectrum
 * 
 * Implements class and makes fields no longer optional.
 */
export interface ColorSpectrum extends ColorSpectrumList {
   blue: string;
   indigo: string;
   purple: string;
   pink: string;
   red: string;
   orange: string;
   yellow: string;
   green: string;
   teal: string;
   cyan: string;
   /**
    * Color for the owner of message/conversation
    */
   owner: string;
}
/**
 * Colors found on the theme
 */
export interface ThemeColors extends ColorSpectrum {
   primary: string;
   secondary: string;
   success: string;
   warning: string;
   danger: string;
   info: string;
   dark: string;
   light: string;
   text: string;
   white: string;
   gray: string;
   grayDark: string;
}
/**
 * Possible colors available for use in theme
 */
export type ThemeColor = keyof ThemeColors;
/**
 * Background styles found on the theme
 */
export interface ThemeBackgrounds {
   page: string;
   message: string;
   buttonPrimary: string;
   inputField: string;
   conversation: string;
   conversationSelected: string;
}

/**
 * Export of Theme to 'styled-components'
 */
declare module 'styled-components' {
   /**
    * Theme interface
    */
   export interface Theme {
      color: ThemeColors;
      background: ThemeBackgrounds;
      border: string;
   }
   /**
    * Possible colors available for use in theme
    */
   export type ThemeColor = keyof ThemeColors;
}
/**
 * Theme types
 */
export type ThemeName = 'dark' | 'light';

/**
 * Checks if given value is a valid theme name.
 *
 * Returns `true` if given value is a valid theme name.
 *
 * Returns `false` if given value is not a valid theme name.
 * @param value Any given value to check if matches theme names
 */
export function isThemeName(value: any): value is ThemeName {
   if (value === 'light' || value === 'dark') {
      return true;
   }
   return false;
}
