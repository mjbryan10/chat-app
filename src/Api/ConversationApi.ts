import BaseApi from './BaseApi';
import axios, { AxiosResponse } from 'axios';
import { Conversation, ApiPostResponse } from './@types';

export default class ConversationApi extends BaseApi {
   async fetchConversationsByUserId(
      userId: number
   ): Promise<AxiosResponse<Conversation[]>> {
      const response = await axios
         .get(`${this.baseUrl}conversation/user/${userId}`)
         .catch((error) => {
            throw error;
         });
      return response;
   }

   async createNewPersonalConversation(
      users: number[]
   ): Promise<AxiosResponse<ApiPostResponse>> {
      const response = await axios
         .post(`${this.baseUrl}conversation/personal`, {
            users,
         })
         .catch((error) => {
            throw error;
         });

      return response;
   }

   async createNewGroupConversation(
      users: number[],
      name: string
   ): Promise<AxiosResponse<ApiPostResponse>> {
      const response = await axios
         .post(`${this.baseUrl}conversation/group`, {
            users,
            name,
         })
         .catch((error) => {
            throw error;
         });

      return response;
   }
}
