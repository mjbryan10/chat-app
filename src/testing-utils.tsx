import React from 'react';
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { dark } from './shared/theme';

const AllTheProviders = ({ children }: any) => {
  return (
    <ThemeProvider theme={dark}>
        {children}
    </ThemeProvider>
  )
  
}

const customRender = (ui?: any, options?: any) =>
  render(ui, { wrapper: AllTheProviders, ...options })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }