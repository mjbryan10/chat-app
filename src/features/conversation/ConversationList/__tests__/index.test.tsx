import React from 'react';
import { render } from 'testing-utils';
import ConversationList from '..';
import { Conversation } from 'shared/Api/types';

describe('ConversationList Feature Component', () => {
   const __privateConversation: Conversation = {
      conversation: {
         id: 4010,
         conversationId: 4010,
         is_owner: 0,
         userid: 1,
         status: 1,
         lastseen: null,
         name: null,
         type: 1,
      },
      users: [
         {
            id: 10257,
            conversationId: 4010,
            is_owner: 0,
            userid: 1,
            status: null,
            lastseen: null,
         },
         {
            id: 10258,
            conversationId: 4010,
            is_owner: 0,
            userid: 2,
            status: null,
            lastseen: null,
         },
      ],
   };
   const __groupConversation: Conversation = {
      conversation: {
         id: 4,
         conversationId: 4,
         is_owner: 0,
         userid: 1,
         status: 1,
         lastseen: '2019-11-05 10:10:50',
         name: 'Group Chaaat',
         type: 2,
      },
      users: [
         {
            id: 31,
            conversationId: 4,
            is_owner: 0,
            userid: 1,
            status: null,
            lastseen: '2019-11-05 10:10:50',
         },
         {
            id: 32,
            conversationId: 4,
            is_owner: 0,
            userid: 2,
            status: null,
            lastseen: '2019-01-25 11:49:01',
         },
         {
            id: 33,
            conversationId: 4,
            is_owner: 0,
            userid: 3,
            status: null,
            lastseen: null,
         },
         {
            id: 34,
            conversationId: 4,
            is_owner: 0,
            userid: 4,
            status: null,
            lastseen: null,
         },
      ],
   };

   const __conversations = [__privateConversation, __groupConversation];
   test('should render correctly', () => {
      const { asFragment } = render(<ConversationList conversations={__conversations} />);

      expect(asFragment()).toMatchSnapshot();
   });
});
