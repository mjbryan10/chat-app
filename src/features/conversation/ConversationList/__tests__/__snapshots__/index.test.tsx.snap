// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConversationList Feature Component should render correctly 1`] = `
<DocumentFragment>
  <button
    class="sc-AxjAm jJUWfx"
    color="blue"
    data-testid="conversation-item-button"
  >
    <div
      class="sc-AxirZ uACJY"
      color="blue"
    >
      <div
        class="sc-AxiKw cefOCi"
        color="blue"
        size="50"
      >
        <span
          class="sc-AxhCb jPjWba"
          data-testid="avatar-inner-container"
          size="50"
        >
          L
        </span>
      </div>
      <h3
        data-testid="avatar-with-label-span"
      >
        Loading..
      </h3>
    </div>
  </button>
  <button
    class="sc-AxjAm kpMVKU"
    color="red"
    data-testid="conversation-item-button"
  >
    <div
      class="sc-AxirZ knpAQZ"
      color="red"
    >
      <div
        class="sc-AxiKw gVga-dC"
        color="red"
        size="50"
      >
        <span
          class="sc-AxhCb jPjWba"
          data-testid="avatar-inner-container"
          size="50"
        >
          L
        </span>
      </div>
      <h3
        data-testid="avatar-with-label-span"
      >
        Loading..
      </h3>
    </div>
  </button>
</DocumentFragment>
`;
